inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/gather_nd_ref.cl

```diff
@@ -0,0 +1,225 @@
+    #else
+        uint pitch_acc = 1;
+        uint output_batch_size = 0;
+        for (int i = BATCH_DIMS - 1; i >= 0; i--) {
+            output_batch_size += (idx_arr[i] * pitch_acc);
+            pitch_acc *= idx_dim[i];
+        }
+
+        #if OUTPUT_DIMS == 4
+            const uint out_x = idx_arr[BATCH_DIMS+2];
+            const uint out_y = idx_arr[BATCH_DIMS+1];
+        #elif OUTPUT_DIMS == 5
+            const uint out_x = idx_arr[BATCH_DIMS+3];
+            const uint out_y = idx_arr[BATCH_DIMS+2];
+            const uint out_z = idx_arr[BATCH_DIMS+1];
+        #elif OUTPUT_DIMS == 6
+            const uint out_x = idx_arr[BATCH_DIMS+4];
+            const uint out_y = idx_arr[BATCH_DIMS+3];
+            const uint out_z = idx_arr[BATCH_DIMS+2];
+            const uint out_w = idx_arr[BATCH_DIMS+1];
+        #endif
+        const uint out_f = idx_arr[BATCH_DIMS+0];
+        const uint out_b = output_batch_size;
+    #endif
+
+    const uint output_idx = GET_OUTPUT_INDEX(OUT_ORDER);
+
+    // Copy data to output as slice size
+    #if HAS_FUSED_OPS
+        #if OUTPUT_DIMS == 4
+            const uint y_pitch = OUTPUT_SIZE_X;
+            const uint f_pitch = y_pitch * OUTPUT_SIZE_Y;
+        #elif OUTPUT_DIMS == 5
+            const uint y_pitch = OUTPUT_SIZE_X;
+            const uint z_pitch = y_pitch * OUTPUT_SIZE_Y;
+            const uint f_pitch = z_pitch * OUTPUT_SIZE_Z;
+        #elif OUTPUT_DIMS == 6
+            const uint y_pitch = OUTPUT_SIZE_X;
+            const uint z_pitch = y_pitch * OUTPUT_SIZE_Y;
+            const uint w_pitch = z_pitch * OUTPUT_SIZE_Z;
+            const uint f_pitch = w_pitch * OUTPUT_SIZE_W;
+        #endif
+        const uint b_pitch = f_pitch * OUTPUT_FEATURE_NUM;
+    #endif
+
+    for (int i = 0; i < WI_SLICE_SIZE; i++) {
+        uint dst_idx = output_idx + i;
+        INPUT0_TYPE val = data[data_idx + i];
+
+        #if HAS_FUSED_OPS
+            const uint b_remain = dst_idx % b_pitch;
+            const uint f_remain = b_remain % f_pitch;
+            #if OUTPUT_DIMS == 4
+                const uint y_remain = f_remain % y_pitch;
+
+                const uint y = f_remain / y_pitch;
+            #elif OUTPUT_DIMS == 5
+                const uint z_remain = f_remain % z_pitch;
+                const uint y_remain = z_remain % y_pitch;
+
+                const uint z = f_remain / z_pitch;
+                const uint y = z_remain / y_pitch;
+            #elif OUTPUT_DIMS == 6
+                const uint w_remain = f_remain % w_pitch;
+                const uint z_remain = w_remain % z_pitch;
+                const uint y_remain = z_remain % y_pitch;
+
+                const uint w = f_remain / w_pitch;
+                const uint z = w_remain / z_pitch;
+                const uint y = z_remain / y_pitch;
+            #endif
+            const uint b = dst_idx / b_pitch;
+            const uint f = b_remain / f_pitch;
+            const uint x = y_remain;
+
+            FUSED_OPS;
```

Comment:
Maybe it makes sence to try use FUSED_OPS_PRELOAD and FUSED_OPS_CALC macro for getting better perf values?

Reply:
Done.